File: lib/outlier.py
from scipy.stats import norm
import numpy as np

# Compute the z-score, and tail probability outlier score.
#     zscore(x) = |x - EX| / σ(X)
#     outlier score (x) = -log P(|X - EX| >= |x - EX|)
class ZOutlierScore:
    def __init__(self, X):
        self.loc = np.mean(X) 
        self.scale = np.std(X) 

    def score(self, X):
        return -norm.logcdf(np.abs((X - self.loc) / self.scale))

def test_z_scorer():
    X = np.random.rand(10)
    scorer = ZOutlierScore(X)
    scorer.loc
    scorer.scale
    scorer.score(X)

File: lib/outlier.jl
using Statistics, Distributions
using Parameters: @unpack

""" Compute the z-score, and tail probability outlier score.
    zscore(x) = |x - EX| / σ(X)
    outlier score (x) = -log P(|X - EX| >= |x - EX|)
"""
struct ZOutlierScore
    μ
    σ
end

function Distributions.fit(ZOutlierScore, X)
    μ, σ = mean(X), std(X)
    ZOutlierScore(μ, σ)
end

function (score::ZOutlierScore)(X)
    @unpack μ, σ = score
    standard_normal = Normal(0, 1)
    @. -logcdf(standard_normal, -abs(X - μ) / σ)
end

